File: exec/monitor/ownGateway.py
#!/usr/bin/python


# << ownGateway v.1.0>>
# Hashed password calculator for BTicino gateways


#   M O D U L E S  #

import ConfigParser
import socket


#   V A R I A B L E S   #

# Acknowledge
ACK='*#*1##'
# Monitor session
MONITOR='*99*1##'
# Commands session
COMMANDS='*99*0##'


#   F U N C T I O N S   #

def Connect(gwMode, first_time = True):
    # Read config from ownGateway.conf
    config = ConfigParser.RawConfigParser()
    config.read('/var/www/exec/ownGateway.conf')
    gwHost = config.get('GATEWAY', 'HOST');
    gwPort = int(config.get('GATEWAY', 'PORT'));
    gwPassword = config.get('GATEWAY', 'PASSWORD');
    # Connect to gateway
    can_return = first_time
    if not first_time:
        pass
        #print 'Start reconnect loop...'
        #WriteLog('Start reconnect loop...')
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    while 1:
        try:
            sock.connect((gwHost, gwPort))
            # Check the gateway answer
            if sock.recv(1024) != ACK:
                raise Exception('Did not receive expected ACK (1)!')
            # It's OK, open CMD or MON mode
            sock.send(gwMode)
            # If password is present, execute 'nonce-hash' authentication
            if gwPassword != '':
                # Receive 'Nonce' from gateway
                Nonce = sock.recv(1024)
                # Calculate hashed pwd
                hpwd = '*#' + str(CalcPass(gwPassword, Nonce)) + '##'
                # Send hashed pwd
                sock.send(hpwd)
                # Check the gateway answer
                if sock.recv(1024) != ACK:
                    raise Exception('Did not receive expected ACK (2)!')
            can_return = True
        except socket.error as e:
            #print 'Socket exception raised: n.{0}: {1}'.format(e.errno, e.strerror)
            #WriteLog('Socket exception raised: n.{0}: {1}'.format(e.errno, e.strerror))
            sock.close()
            if first_time:
                sock = None
        except Exception as e:
            #print 'Unexpected exception raised: {0}'.format(e.message)
            #WriteLog('Unexpected exception raised: {0}'.format(e.message))
            sock.close()
            if first_time:
                sock = None
        if can_return:
            if not first_time:
                pass
                #print 'Gateway connection restored!'
                #WriteLog('Gateway connection restored!')
            return sock
        time.sleep(1)


def CalcPass(password, nonce):
    m_1 = 0xFFFFFFFFL
    m_8 = 0xFFFFFFF8L
    m_16 = 0xFFFFFFF0L
    m_128 = 0xFFFFFF80L
    m_16777216 = 0XFF000000L
    flag = True
    num1 = 0L
    num2 = 0L
    password = long(password)
 
    for c in nonce:
        num1 = num1 & m_1
        num2 = num2 & m_1
        if c == '1':
            length = not flag
            if not length:
                num2 = password
            num1 = num2 & m_128
            num1 = num1 >> 7
            num2 = num2 << 25
            num1 = num1 + num2
            flag = False
        elif c == '2':
            length = not flag
            if not length:
                num2 = password
            num1 = num2 & m_16
            num1 = num1 >> 4
            num2 = num2 << 28
            num1 = num1 + num2
            flag = False
        elif c == '3':
            length = not flag
            if not length:
                num2 = password
            num1 = num2 & m_8
            num1 = num1 >> 3
            num2 = num2 << 29
            num1 = num1 + num2
            flag = False
        elif c == '4':
            length = not flag
 
            if not length:
                num2 = password
            num1 = num2 << 1
            num2 = num2 >> 31
            num1 = num1 + num2
            flag = False
        elif c == '5':
            length = not flag
            if not length:
                num2 = password
            num1 = num2 << 5
            num2 = num2 >> 27
            num1 = num1 + num2
            flag = False
        elif c == '6':
            length = not flag
            if not length:
                num2 = password
            num1 = num2 << 12
            num2 = num2 >> 20
            num1 = num1 + num2
            flag = False
        elif c == '7':
            length = not flag
            if not length:
                num2 = password
            num1 = num2 & 0xFF00L
            num1 = num1 + (( num2 & 0xFFL ) << 24 )
            num1 = num1 + (( num2 & 0xFF0000L ) >> 16 )
            num2 = ( num2 & m_16777216 ) >> 8
            num1 = num1 + num2
            flag = False
        elif c == '8':
            length = not flag
            if not length:
                num2 = password
            num1 = num2 & 0xFFFFL
            num1 = num1 << 16
            num1 = num1 + ( num2 >> 24 )
            num2 = num2 & 0xFF0000L
            num2 = num2 >> 8
            num1 = num1 + num2
            flag = False
        elif c == '9':
            length = not flag
            if not length:
                num2 = password
            num1 = ~num2
            flag = False
        else:
            num1 = num2
        num2 = num1
    return num1 & m_1


def Disconnect(sock):
    # Disconnect from Gateway
    sock.close()


def RecData(sock):
    # Read data from Gateway
    return sock.recv(1024)


def SendData(sock, cmdopen):
    # Send data to Gateway
    sock.send(cmdopen)

File: exec/ownGateway.php
<?php

define('ACK', '*#*1##');       // Acknowledge
define('MONITOR', '*99*1##');  // Monitor session
define('COMMANDS', '*99*0##'); // Commands session

function ownConnect($gwMode) {
    // Read config from ownGateway.conf
    $config = parse_ini_file(dirname(__FILE__).'/ownGateway.conf');
    $gwHost = $config['HOST'];
    $gwPort = $config['PORT'];
    $gwPassword = $config['PASSWORD'];
    // Connect to gateway
    $sock = fsockopen($gwHost, $gwPort);
    if ($sock) {
        try {
            $data = fread($sock, 1024);
            if ($data != ACK) {
                throw new Exception('Did not receive expected ACK (1)!');
            }
            // It's OK, open Command or Monitor mode
            fwrite($sock, $gwMode);
            // If password is present, execute "nonce-hash" authentication
            if ($gwPassword != '') {
                // Receive "Nonce" from gateway
                $data = fread($sock, 1024);
                // Calculate hashed pwd
                $data = '*#'.ownCalcPass($gwPassword, $data).'##';
                // Send hashed pwd
                fwrite($sock, $data);
                $data = fread($sock, 1024);
                if ($data != ACK) {
                    throw new Exception('Did not receive expected ACK (2)!');
                }
            }
        } catch (Exception $e) {
            fclose($sock);
            $sock = False;
        }
    }
    return $sock;
}

function ownCalcPass($password, $nonce) {
    $msr = 0x7FFFFFFF;
    $m_1 = (int)0xFFFFFFFF;
    $m_8 = (int)0xFFFFFFF8;
    $m_16 = (int)0xFFFFFFF0;
    $m_128 = (int)0xFFFFFF80;
    $m_16777216 = (int)0xFF000000;
    $flag = True;
    $num1 = 0;
    $num2 = 0;
    foreach (str_split($nonce) as $c) {
        $num1 = $num1 & $m_1;
        $num2 = $num2 & $m_1;
        if ($c == '1') {
            $length = !$flag;
            if (!$length) {
                $num2 = $password;
            }
            $num1 = $num2 & $m_128;
            $num1 = $num1 >> 1;
            $num1 = $num1 & $msr;
            $num1 = $num1 >> 6;
            $num2 = $num2 << 25;
            $num1 = $num1 + $num2;
            $flag = False;
        } elseif ($c == '2') {
            $length = !$flag;
            if (!$length) {
                $num2 = $password;
            }
            $num1 = $num2 & $m_16;
            $num1 = $num1 >> 1;
            $num1 = $num1 & $msr;
            $num1 = $num1 >> 3;
            $num2 = $num2 << 28;
            $num1 = $num1 + $num2;
            $flag = False;
        } elseif ($c == '3') {
            $length = !$flag;
            if (!$length) {
                $num2 = $password;
            }
            $num1 = $num2 & $m_8;
            $num1 = $num1 >> 1;
            $num1 = $num1 & $msr;
            $num1 = $num1 >> 2;
            $num2 = $num2 << 29;
            $num1 = $num1 + $num2;
            $flag = False;
        } elseif ($c == '4') {
            $length = !$flag;
            if (!$length) {
                $num2 = $password;
            }
            $num1 = $num2 << 1;
            $num2 = $num2 >> 1;
            $num2 = $num2 & $msr;
            $num2 = $num2 >> 30;
            $num1 = $num1 + $num2;
            $flag = False;
        } elseif ($c == '5') {
            $length = !$flag;
            if (!$length) {
                $num2 = $password;
            }
            $num1 = $num2 << 5;
            $num2 = $num2 >> 1;
            $num2 = $num2 & $msr;
            $num2 = $num2 >> 26;
            $num1 = $num1 + $num2;
            $flag = False;
        } elseif ($c == '6') {
            $length = !$flag;
            if (!$length) {
                $num2 = $password;
            }
            $num1 = $num2 << 12;
            $num2 = $num2 >> 1;
            $num2 = $num2 & $msr;
            $num2 = $num2 >> 19;
            $num1 = $num1 + $num2;
            $flag = False;
        } elseif ($c == '7') {
            $length = !$flag;
            if (!$length) {
                $num2 = $password;
            }
            $num1 = $num2 & 0xFF00;
            $num1 = $num1 + (( $num2 & 0xFF ) << 24 );
            $num1 = $num1 + (( $num2 & 0xFF0000 ) >> 16 );
            $num2 = $num2 & $m_16777216;
            $num2 = $num2 >> 1;
            $num2 = $num2 & $msr;
            $num2 = $num2 >> 7;
            $num1 = $num1 + $num2;
            $flag = False;
        } elseif ($c == '8') {
            $length = !$flag;
            if (!$length) {
                $num2 = $password;
            }
            $num1 = $num2 & 0xFFFF;
            $num1 = $num1 << 16;
            $numx = $num2 >> 1;
            $numx = $numx & $msr;
            $numx = $numx >> 23;
            $num1 = $num1 + $numx;
            $num2 = $num2 & 0xFF0000;
            $num2 = $num2 >> 1;
            $num2 = $num2 & $msr;
            $num2 = $num2 >> 7;
            $num1 = $num1 + $num2;
            $flag = False;
        } elseif ($c == '9') {
            $length = !$flag;
            if (!$length) {
                $num2 = $password;
            }
            $num1 = ~(int)$num2;
            $flag = False;
        } else {
            $num1 = $num2;
        }
        $num2 = $num1;
    }
    return sprintf('%u', $num1 & $m_1);
}

?>
